src/wp-includes/html-api/class-wp-html-tag-processor.php

```diff
@@ -2281,6 +2590,171 @@ public function is_tag_closer() {
+		switch ( $this->parser_state ) {
+			case self::STATE_MATCHED_TAG:
+				return $this->get_tag();
+
+			case self::STATE_TEXT_NODE:
+				return '#text';
+
+			case self::STATE_CDATA_NODE:
+				return '#cdata-section';
+
+			case self::STATE_PI_NODE:
+				return substr( $this->html, $this->tag_name_starts_at, $this->tag_name_length );
+
+			case self::STATE_COMMENT:
+				return '#comment';
+
+			case self::STATE_DOCTYPE:
+				return 'html';
+
+			case self::STATE_PRESUMPTUOUS_TAG:
+				return '#presumptuous-tag';
+
+			case self::STATE_FUNKY_COMMENT:
+				return '#funky-comment';
+		}
+	}
+
+	/**
+	 * Returns the modifiable text for a matched token, or an empty string.
+	 *
+	 * Modifiable text is text content that may be read and changed without
+	 * changing the HTML structure of the document around it. This includes
+	 * the contents of `#text` nodes in the HTML as well as the inner
+	 * contents of HTML comments, Processing Instructions, and others, even
+	 * though these nodes aren't part of a parsed DOM tree. They also contain
+	 * the contents of SCRIPT and STYLE tags, of TEXTAREA tags, and of any
+	 * other section in an HTML document which cannot contain HTML markup (DATA).
+	 *
+	 * If a token has no modifiable text then an empty string is returned to
+	 * avoid needless crashing or type errors. An empty string does not mean
+	 * that a token has modifiable text, and a token with modifiable text may
+	 * have an empty string (e.g. a comment with no contents).
+	 *
+	 * @return string
+	 */
+	public function get_modifiable_text() {
+		if ( null === $this->text_starts_at ) {
+			return '';
+		}
+
+		$text = substr( $this->html, $this->text_starts_at, $this->text_length );
+
+		// Comment data is not decoded.
+		if (
+			self::STATE_CDATA_NODE === $this->parser_state ||
+			self::STATE_COMMENT === $this->parser_state ||
+			self::STATE_DOCTYPE === $this->parser_state ||
+			self::STATE_PI_NODE === $this->parser_state ||
+			self::STATE_FUNKY_COMMENT === $this->parser_state
+		) {
+			return $text;
+		}
+
+		$tag_name = $this->get_tag();
+		if (
+			// Script data is not decoded.
+			'SCRIPT' === $tag_name ||
+
+			// RAWTEXT data is not decoded.
+			'IFRAME' === $tag_name ||
+			'NOEMBED' === $tag_name ||
+			'NOFRAMES' === $tag_name ||
+			'STYLE' === $tag_name ||
+			'XMP' === $tag_name
+		) {
+			return $text;
+		}
+
+		$decoded = html_entity_decode( $text, ENT_QUOTES | ENT_HTML5 | ENT_SUBSTITUTE );
+
+		if ( empty( $decoded ) ) {
+			return '';
+		}
+
+		/*
+		 * PRE and TEXTAREA skip a leading newline, but this newline may not be a `\n`.
+		 * It could be a character reference, such as `<textarea>&#x0a;Content</textarea>`.
+		 * For these cases it's important to first decode the text content before checking
+		 * for a leading newline and removing it.
+		 */
+		if ( self::STATE_MATCHED_TAG === $this->parser_state ) {
+			switch ( $tag_name ) {
+				case 'PRE':
+				case 'TEXTAREA':
+					if ( "\n" === $decoded[0] ) {
```

Comment:
[This should account for `LISTING` too](https://software.hixie.ch/utilities/js/live-dom-viewer/?%3Cbody%3E%3Cpre%3E%26%2310%3B%3C%2Fpre%3E%3Clisting%3E%26%23xa%3B%3C%2Flisting%3E%3Ctextarea%3E%0A%3C%2Ftextarea%3E%3Cdiv%3E%0A%3C%2Fdiv%3E):

> A start tag whose tag name is one of: "pre", "listing"
> …[Insert an HTML element](https://html.spec.whatwg.org/multipage/parsing.html#insert-an-html-element) for the token.
> …
> If the [next token](https://html.spec.whatwg.org/multipage/parsing.html#next-token) is a U+000A LINE FEED (LF) character token, then ignore that token and move on to the next one. (Newlines at the start of [pre](https://html.spec.whatwg.org/multipage/grouping-content.html#the-pre-element) blocks are ignored as an authoring convenience.)

```suggestion
		/*
		 * LISTING, PRE, and TEXTAREA skip a leading newline, but this newline may not be a `\n`.
		 * It could be a character reference, such as `<textarea>&#x0a;Content</textarea>`.
		 * For these cases it's important to first decode the text content before checking
		 * for a leading newline and removing it.
		 */
		if ( self::STATE_MATCHED_TAG === $this->parser_state ) {
			switch ( $tag_name ) {
				case 'LISTING':
				case 'PRE':
				case 'TEXTAREA':
					if ( "\n" === $decoded[0] ) {
```

Reply:
accepted and merged